academy/0.0-B9lab-Blockchains/5_crypto.md

```diff
@@ -0,0 +1,186 @@
+
+1. Alice encrypts her message with Bob's public key.
+2. Alice signs the encrypted file with her private key.
+3. Upon reception, Bob verifies the signature with Alice's public key to make sure the file came from Alice.
+4. Bob decrypts the file with his private key.
+
+![Encryption, decryption, and signatures with public/private keys](images/00_13_mix_n_match_keys_v3.png)
+
+### Is this science or magic?
+
+If these examples seem counter-intuitive it means you sense the mathematical shadow of public-key encryption. You do not have to understand the mathematics behind it at a deep level, but you must understand the properties and implications of using public-key cryptography.
+
+Given four keys (A, B, C, and D), we can encrypt a message with three keys (A, B, and C) such that the fourth key (in this case, D) is required to decrypt it and is very hard to guess or discover. So, if Alice knows her private key and her public key and she also knows Bob's public key, she can encrypt a message that can only be understood by someone who knows Bob's private key.
+
+Similarly, given knowledge of public and private keys, one can generate a signature (i.e. a character string) so that someone with a copy of the message and the signature can independently determine the public key of the entity that signed the message, proving that the signer knows the corresponding private key.
+
+It is then possible to proceed with the understanding that signed messages from Alice could only come from someone with knowledge of Alice's private key, and messages that are encrypted for Bob can only be deciphered by Bob.
+
+### Key management and public key infrastructure
+
+If you look again at the Alice and Bob examples, you will notice that there is a vulnerability in "Bob gives Alice his public key". A malicious Charlie could intercept Bob's public key and pass on _his own_ public key to Alice.
+
+Key management and public key infrastructure (PKI) is an important aspect of cryptography that helps mitigate this risk.
+
+## Cryptographic hash functions
+
+Blockchain technology relies heavily on hash functions, as they help establish the so-called "chain of blocks".
+
+All cryptographic hash functions fulfill several properties:
+
+* Converts an input (a.k.a. the message) into an output (a.k.a the hash).
+* Does the conversion in a reasonable amount of time.
+* It is practically impossible to re-generate the message out of the hash.
+* The tiniest change in the message changes the hash beyond recognition.
+* It is practically impossible to find two different messages with the same hash.
+
+With such a function, you can:
+
+* Prove that you have a message without disclosing the content of the message, for instance:
+    * To prove you know your password.
+    * To prove you previously wrote a message.
+* Be confident that the message was not altered.
+* Index your messages.
+
+### A closer look at a hash function
+
+MD5 is such a hash function:
```

Comment:
It is better to not use a hash function that has been proven to not fulfill the description of above.